### ISSUE TITLE
Add MemoryMappedDataSetIterator

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/nd4j#1521]
Idea: Things like shuffling are different with presaved datasets. Caching is also an issue.
Often when we are doing gradient descent we need to do shuffling among other things.

If we had a memory mapped file, we could identify records at offsets and index in to 1 memory mapped file for assembling datasets on the fly.

This would solve a few things: a class balance problem we often run in to, shuffling in near constant time, and sampling.

**Comment by agibsonccc** [comment]
Possibly use a MemoryMappedOutputStream http://stackoverflow.com/questions/20495278/memory-mapped-output-stream-produces-an-output-file-with-trailing-zeros in addition to dataset.save/load?

We will still need to track offsets here.

**Comment by AlexDBlack** [comment]
I very much like the general strategy of using a memory mapped file to back iterators.

Some issues that this approach may solve:
- Avoid parsing data multiple times (i.e., parse CSV *every* epoch, read and resize images multiple times, etc)
- Shuffling between epochs (not currently possible/easy with CSV, for example)
    - Shuffling between epochs: has a performance (accuracy) hit, but is hard to do with current functionality
    - Implementing shuffling at the RecordReader level: end up writing the same functionality over and over...
- Simplifies data set preprocessing/persistence in binary form - i.e., reuse the MM file contents multiple times
- Solves the 'how many examples' and 'how many DataSets' issues (i.e., export first -> known size)


Some design thoughts and issues:
- The data for each example should be stored separately, not batched. Internally, could just be INDArrays. Storing batches should be possible, but has some downsides:
    - Less efficient reads of arbitrary examples (i.e., shuffling), unless strides are exactly right for contiguous memory access
    - Less efficient storage, in the case of time series (need mask arrays, and store masked values)
- Could store data in compressed format (to minimize disk space, total disk read, number of page misses)
- We should provide persisted and non-persisted options
    - Sometimes reuse of the data is desirable, sometimes we don't (manually having to delete a file in the latter case isn't ideal)
    - When persisted: allow people to load and use it again easily
- Current metadata functionality should be considered here too (i.e., store metadata also when present)
    - Say I have a test set: evaluation from MM file -> which examples were classified incorrectly?
- I don't think this is really usable for HDFS/Spark for exporting (the format is fine, but random HDFS reads is the problem). Unless we can find a workaround? (Might be better than a million separate DataSet files on HDFS for the export step)

	
Some thoughts on how this could work: 2 implementations
- One that replaces RecordReaderDataSetIterator -> load and persist one example at a time from a RecordReader
- One that simply wraps other DataSetIterators -> same underlying implementation as above, but this one does the splitting into separate examples first, if necessary

**Comment by AlexDBlack** [comment]
Side note: could we use the same underlying storage format for Keras?
Suppose your training data is in python -> export to file -> read in DL4J as MM file

**Comment by pkoperek** [comment]
@AlexDBlack Yup that would simplify things - currently I'm just dumping data to HDF5 format just before starting learning and send the path of the file to dl4j. Hacky test script for reference: https://github.com/pkoperek/keras-dl4j/blob/master/mnist_example_full.py

**Comment by treo** [comment]
Another thing that a memory mapped file has going for it is that while still being off-heap, the data doesn't have to be moved over to cuda if it isn't actually used.

I used that for word vector loading when I found the default way to be too slow (see this 4 month old gist: https://gist.github.com/treo/f5a346d53f89566b51bf88a9a42c67c7Z). Only one thing you should notice: While the file can be larger than 2GB, you can't actually index directly into it without splitting it into chunks as Java doesn't allow integers that are that large.

**Comment by AlexDBlack** [comment]
Additional thought: having the ability to use this to perform sampling (in the case of imbalanced data sets - over-sample infrequent classes, under-sample frequent classes) would be a nice addition.
